### ISSUE TITLE
Products API paramater issue

### ISSUE THREAD

**Issue by ImprobableCode** [woothemes/woocommerce#35719]
### Prerequisites

- [X] I have carried out troubleshooting steps and I believe I have found a bug.
- [X] I have searched for similar bugs in both open and closed issues and cannot find a duplicate.

### Describe the bug

When sending a GET request to the API, adding the param catalog_visibility to the URL with it set to visible or hidden does not change the returned data. 

The new store API does not exhibit this behaviour

### Expected behavior

Returned data should contain only hidden or visible products, not both

### Actual behavior

returns all products regardless of the flag

### Steps to reproduce

fresh install of the latest woocommerce. 

Add a product with Shop and search results (in the top right publish section)
Add another product but with Hidden 
Setup API and keys in menu,  woocommerce> Settings> Advanced>REST API

Accees the products API through Postman , thunder client or from with WordPress with a fetch or ajax.

https://your_domain.com/wp-json/wc/v3/products?catalog_visibility=visible

### WordPress Environment

Woocommerce Version: WooCommerce 7.2.0
WordPress version: 6.1.1

PC:
Windows 10, 
Chrome(Version 107.0.5304.107)
Firefox(Version 107.0)
Safari (Version 16.1)

### Isolating the problem

- [X] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [x] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/storefront/).
- [X] I can reproduce this bug consistently using the steps above.

**Comment by jonathansadowski** [comment]
Hi @ImprobableCode, 

Thanks for the report. Your report was missing the system status report. We need the system status report in order to try and reproduce the issue. Please reply to this message with the status report.

You can find the report from **WooCommerce > System Status** in WordPress admin. Click on "Get system report" and then "Copy for support".

**Comment by ImprobableCode** [comment]
Sorry about that 
### WordPress Environment ###

WordPress address (URL): https://example.com
Site address (URL): https://example.com
WC Version: 7.1.0
REST API Version: ✔ 7.1.0
WC Blocks Version: ✔ 8.7.5
Action Scheduler Version: ✔ 3.4.0
Log Directory Writable: ✔
WP Version: 6.1.1
WP Multisite: –
WP Memory Limit: 256 MB
WP Debug Mode: –
WP Cron: ✔
Language: en_GB
External object cache: –

### Server Environment ###

Server Info: Apache/2.4.54 (Debian)
PHP Version: 7.4.30
PHP Post Max Size: 100 MB
PHP Time Limit: 30
PHP Max Input Vars: 1000
cURL Version: 7.74.0
OpenSSL/1.1.1n

SUHOSIN Installed: –
MySQL Version: 5.5.5-10.5.15-MariaDB-0+deb11u1
Max Upload Size: 100 MB
Default Timezone is UTC: ✔
fsockopen/cURL: ✔
SoapClient: ❌ Your server does not have the SoapClient class enabled - some gateway plugins which use SOAP may not work as expected.
DOMDocument: ✔
GZip: ✔
Multibyte String: ✔
Remote Post: ✔
Remote Get: ✔

### Database ###

WC Database Version: 7.1.0
WC Database Prefix: wp_
Total Database Size: 7.19MB
Database Data Size: 5.45MB
Database Index Size: 1.74MB
wp_woocommerce_sessions: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_api_keys: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_attribute_taxonomies: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_downloadable_product_permissions: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_woocommerce_order_items: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_order_itemmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_tax_rates: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_woocommerce_tax_rate_locations: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_shipping_zones: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_woocommerce_shipping_zone_locations: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_shipping_zone_methods: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_woocommerce_payment_tokens: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_payment_tokenmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_log: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_actions: Data: 0.02MB + Index: 0.11MB + Engine InnoDB
wp_actionscheduler_claims: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_groups: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_logs: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_aws_cache: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_aws_index: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_commentmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_comments: Data: 0.02MB + Index: 0.09MB + Engine InnoDB
wp_frm_fields: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_frm_forms: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_frm_items: Data: 0.02MB + Index: 0.08MB + Engine InnoDB
wp_frm_item_metas: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_links: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_options: Data: 4.03MB + Index: 0.06MB + Engine InnoDB
wp_postmeta: Data: 0.19MB + Index: 0.11MB + Engine InnoDB
wp_posts: Data: 0.22MB + Index: 0.06MB + Engine InnoDB
wp_termmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_terms: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_term_relationships: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_term_taxonomy: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_usermeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_users: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_wc_admin_notes: Data: 0.05MB + Index: 0.00MB + Engine InnoDB
wp_wc_admin_note_actions: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_category_lookup: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wc_customer_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_download_log: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_order_coupon_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_order_product_lookup: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_wc_order_stats: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_wc_order_tax_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_product_attributes_lookup: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_product_download_directories: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_product_meta_lookup: Data: 0.02MB + Index: 0.09MB + Engine InnoDB
wp_wc_rate_limits: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_reserved_stock: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wc_tax_rate_classes: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_webhooks: Data: 0.02MB + Index: 0.02MB + Engine InnoDB

### Post Type Counts ###

attachment: 1569
custom_css: 4
customize_changeset: 47
frm_form_actions: 1
frm_styles: 1
gp_elements: 15
nav_menu_item: 26
page: 21
post: 8
product: 1343
revision: 25
shop_coupon: 1
shop_order: 2
wp_global_styles: 2
wp_show_posts: 2

### Security ###

Secure connection (HTTPS): ✔
Hide errors from visitors: ✔

### Active Plugins (6) ###

Advanced Woo Search: by ILLID – 2.68
GenerateBlocks Pro: by Tom Usborne – 1.4.0
GenerateBlocks: by Tom Usborne – 1.6.0
GP Premium: by Tom Usborne – 2.2.1
WooCommerce: by Automattic – 7.1.0
WP Show Posts: by Tom Usborne – 1.1.3

### Inactive Plugins (0) ###


### Settings ###

API Enabled: –
Force SSL: –
Currency: GBP (£)
Currency Position: left
Thousand Separator: ,
Decimal Separator: .
Number of Decimals: 2
Taxonomies: Product Types: external (external)
grouped (grouped)
simple (simple)
variable (variable)

Taxonomies: Product Visibility: exclude-from-catalog (exclude-from-catalog)
exclude-from-search (exclude-from-search)
featured (featured)
outofstock (outofstock)
rated-1 (rated-1)
rated-2 (rated-2)
rated-3 (rated-3)
rated-4 (rated-4)
rated-5 (rated-5)

Connected to WooCommerce.com: –
Enforce Approved Product Download Directories: ✔

### WC Pages ###

Shop base: #6 - /shop/
Basket: #167 - /cart-2/
Checkout: #164 - /checkout-3/
My account: #162 - /account/
Terms and conditions: ❌ Page not set

### Theme ###

Name: blank
Version: 1.0.0
Author URL: 
Child Theme: ✔
Parent Theme Name: GeneratePress
Parent Theme Version: 3.2.4
Parent Theme Author URL: https://tomusborne.com
WooCommerce Support: ✔

### Templates ###

Overrides: –

### Admin ###

Enabled Features: activity-panels
analytics
coupons
customer-effort-score-tracks
experimental-products-task
experimental-import-products-task
experimental-fashion-sample-products
shipping-smart-defaults
shipping-setting-tour
homescreen
marketing
multichannel-marketing
mobile-app-banner
navigation
onboarding
onboarding-tasks
remote-inbox-notifications
remote-free-extensions
payment-gateway-suggestions
shipping-label-banner
subscriptions
store-alerts
transient-notices
woo-mobile-welcome
wc-pay-promotion
wc-pay-welcome-page

Disabled Features: minified-js
new-product-management-experience
settings

Daily Cron: ✔ Next scheduled: 2022-11-29 15:47:32 +00:00
Options: ✔
Notes: 44
Onboarding: completed

### Action Scheduler ###

Complete: 14,304
Oldest: 2022-10-29 06:59:10 +0100
Newest: 2022-11-28 23:48:34 +0000

Pending: 1
Oldest: 2022-11-29 16:03:30 +0000
Newest: 2022-11-29 16:03:30 +0000


### Status report information ###

Generated at: 2022-11-28 23:48:34 +00:00

**Comment by peterfabian** [comment]
Hi, as per the docs, it doesn't seem like this query parameter is supported: https://woocommerce.github.io/woocommerce-rest-api-docs/#list-all-products

I'm changing this into a support request, as there isn't any reason why in this can't work, perhaps there are some performance considerations, though.

**Comment by samrat-biswas-git** [comment]
Thanks @ImprobableCode, and thanks @peterfabian for the earlier steer. To clarify what's happening: catalog_visibility on the products endpoint is a product property (a field on the product object you can set on create/update and read back), not a collection query parameter. It isn't wired into the products query, so GET /products?catalog_visibility=visible doesn't filter — it was never a supported filter, rather than a broken one.

So this is really an enhancement: add catalog_visibility as a filter on the products collection, matching the visibility filtering the Store API already supports.

Marking ready for engineering. It's additive and backwards-compatible (a new opt-in query param). Implementation note: catalog visibility is stored as product_visibility taxonomy terms (exclude-from-catalog, exclude-from-search), so the filter needs a tax_query on those terms rather than a simple column match — the frontend/Store API already do this, so there's a pattern to follow. Worth a test covering visible/hidden/catalog/search.
